### PR TITLE
Add option to load a pretrained model with mismatched shapes

### PR DESCRIPTION
**Pull request by sgugger**
# What does this PR do?

Sometimes, users want to load a checkpoint for a given task with a new head for the same task but different shapes. For instance, they may want to use a checkpoint that does text classification on 2 labels to initialize a model that does text classification on 5 labels. 

This PR enables that by adding a new argument to the `from_pretrained` method of `PreTrainedModel`, `TFPreTrainedModel` and `FlaxPreTrainedModel` named `ignore_mismatched_sizes`. When set to True, this argument will ignore the weights from the checkpoint that do not have the same shape as the ones inside the model and leave the randomly initialized weights.

### CODE REVIEW

File: src/transformers/models/auto/configuration_auto.py [hunk 1]
```diff
@@ -186,7 +186,7 @@
         ("pegasus", PegasusConfig),
         ("marian", MarianConfig),
         ("mbart", MBartConfig),
-        ("megatron_bert", MegatronBertConfig),
+        ("megatron-bert", MegatronBertConfig),
```
Comment (sgugger) [inline src/transformers/models/auto/configuration_auto.py:189]: It looks like a typo since the model type in the config is `"megatron-bert"`. Can change to put both keys if somehow the two handed up being present.

File: src/transformers/models/deberta/modeling_deberta.py [hunk 2]
```diff
@@ -777,25 +773,6 @@ def _init_weights(self, module):
             if module.padding_idx is not None:
                 module.weight.data[module.padding_idx].zero_()
 
-    def _pre_load_hook(self, state_dict, prefix, local_metadata, strict, missing_keys, unexpected_keys, error_msgs):
```
Comment (sgugger) [inline src/transformers/models/deberta/modeling_deberta.py:780]: This logic is not necessary anymore. This is a slightly breaking change though since users will now have to add the `ingore_mismatched_sizes=True` flag to have this work.

Comment (LysandreJik) [reply src/transformers/models/deberta/modeling_deberta.py:780]: That preload hook was set up here in order to do exactly what your PR is doing, so it's fine to remove it here.

File: src/transformers/modeling_tf_utils.py [hunk 3]
```diff
@@ -533,9 +537,14 @@ def load_tf_weights(model, resolved_archive_file, _prefix=None):
                             # If the two shapes are not compatible we raise an issue
                             try:
                                 array = np.reshape(saved_weight_value, K.int_shape(symbolic_weight))
-                            except AssertionError as e:
-                                e.args += (K.int_shape(symbolic_weight), saved_weight_value.shape)
-                                raise e
```
Comment (LysandreJik) [inline src/transformers/modeling_tf_utils.py:538]: Can't that logic be re-used if `not ignore_mismatched_sizes` ?

File: src/transformers/models/auto/configuration_auto.py (hunk 1 above)
Comment (LysandreJik) [reply src/transformers/models/auto/configuration_auto.py:189]: It looks like a typo indeed

File: tests/test_modeling_common.py [hunk 4]
```diff
@@ -1532,6 +1532,34 @@ def test_problem_types(self):
...
+            if model_class not in get_values(MODEL_FOR_SEQUENCE_CLASSIFICATION_MAPPING):
+                continue
+
+            with self.subTest(msg=f"Testing {model_class}"):
+                with tempfile.TemporaryDirectory() as tmp_dir:
+                    model = model_class(config)
+                    model.save_pretrained(tmp_dir)
+
+                    # Fails when we don't set ignore_mismatched_sizes=True
+                    with self.assertRaises(Exception):
```
Comment (LysandreJik) [inline tests/test_modeling_common.py:1548]: Could the exception caught be more explicit than `Exception`?

File: tests/test_modeling_flax_common.py [hunk 5]
```diff
@@ -516,6 +524,32 @@ def test_attention_outputs(self):
...
+            if model_class not in get_values(FLAX_MODEL_FOR_SEQUENCE_CLASSIFICATION_MAPPING):
+                continue
+
+            with self.subTest(msg=f"Testing {model_class}"):
+                with tempfile.TemporaryDirectory() as tmp_dir:
+                    model = model_class(config)
+                    model.save_pretrained(tmp_dir)
+
+                    # Fails when we don't set ignore_mismatched_sizes=True
+                    with self.assertRaises(Exception):
```
Comment (LysandreJik) [inline tests/test_modeling_flax_common.py:540]: Same here

File: tests/test_modeling_tf_common.py [hunk 6]
```diff
@@ -1308,6 +1311,34 @@ def test_generate_with_headmasking(self):
...
+
+            with self.subTest(msg=f"Testing {model_class}"):
+                with tempfile.TemporaryDirectory() as tmp_dir:
+                    model = model_class(config)
+                    inputs = self._prepare_for_class(inputs_dict, model_class)
+                    _ = model(**inputs)
+                    model.save_pretrained(tmp_dir)
+
+                    # Fails when we don't set ignore_mismatched_sizes=True
+                    with self.assertRaises(Exception):
```
Comment (LysandreJik) [inline tests/test_modeling_tf_common.py:1329]: Same here

**Review by LysandreJik — APPROVED** [review]
This looks great, thanks for working on it @sgugger!

File: src/transformers/modeling_tf_utils.py (hunk 3 above)
Comment (sgugger) [reply src/transformers/modeling_tf_utils.py:538]: The error already contains the shape, so I didn't see the point of having it twice.

File: src/transformers/modeling_flax_utils.py [hunk 7]
```diff
@@ -367,6 +372,22 @@ def from_pretrained(
...
+        mismatched_keys = []
+        for key in state.keys():
+            if key in random_state and state[key].shape != random_state[key].shape:
+                if ignore_mismatched_sizes:
+                    mismatched_keys.append((key, state[key].shape, random_state[key].shape))
+                    state[key] = random_state[key]
+                else:
+                    raise ValueError(
+                        f"Trying to load the pretrained weight for {key} failed: checkpoint has shape "
+                        f"{state[key].shape} which is incompatible with the model shape ({random_state[key].shape}). "
```
Comment (patrickvonplaten) [inline src/transformers/modeling_flax_utils.py:386]:
```suggestion
                        f"{state[key].shape} which is incompatible with the model shape {random_state[key].shape}. "
```
Think it's better to not add `(` and `)` so that `{state[key].shape}` and `{random_state[key].shape}` are of the same formard to not confuse the reader

**Review by patrickvonplaten — APPROVED** [review]
